### ISSUE TITLE
Union[NoReturn, A] is handled in unexpected ways

### ISSUE THREAD

**Issue by sorcio** [microsoft/pyright#1919]
In some cases, a value of type `Union[NoReturn, A]` is handled in unexpected ways.

**Function returning  `Union[NoReturn, A]`**

```python
from typing import Union

class A:
    def foo(self):
        return 14

class B:
    def foo(self):
        raise TypeError

C = Union[A, B]

def get_c() -> C:
    return A()

def something() -> int:
    c = get_c()
    value = c.foo()
    return value  # error: Expression of type "NoReturn | Literal[14]" cannot be assigned to return type "int"
```

I would expect pyright to allow the assignment/return, because `value` can only have the same type as `A.foo()` at that point of the program.

**Argument of union type**

With the same type definitions as the above example:

```python
def something2(c: C) -> int:
    value = c.foo()
    return value  # this is considered unreachable
```

I would expect the statements after the call to `c.foo()` to be still considered reachable, because `c` could have type `A`.

**Concrete use case?**

I encountered this issue in the real world while using the [outcome](https://github.com/python-trio/outcome) library:

```python
def fn() -> int:
    ...

...
result = capture(fn)
x: int = result.unwrap()
```

Pyright is going to complain that `result.unwrap()` cannot be assigned to int because it has type `NoResult | int`. 

---

These cases can probably be worked around, but I suspect this might be a bug. It looks similar to https://github.com/microsoft/pyright/issues/1308 but there is no iteration involved here. 

This has been tested with pyright 1.1.144 and Python 3.6.

**Comment by erictraut** [comment]
Thanks for the bug report. I agree that call expressions that evaluate to a union type that includes NoReturn should ignore the NoReturn type. I've updated the type checking logic accordingly.

**Comment by erictraut** [comment]
This is included in pyright 1.1.145, which I just published. It will also be included in the next release of pylance.
